### PR TITLE
adding VSHNRedis ServiceMonitor to expose Redis metrics in Openshift

### PR DESCRIPTION
**Pull request by wejdross**
This is a simple change which:
- adds Redis sidecar with metrics exporter
- adds ServiceMonitor so metrics are available within namespace in Openshift console
- disables securityContext in metrics exporter, so default from Openshift is used

Tested with success on LAB cluster

## Checklist

- [x] The PR has a meaningful title. It will be used to auto generate the
      changelog.
      The PR has a meaningful description that sums up the change. It will be
      linked in the changelog.
- [x] PR contains a single logical change (to build a better changelog).
- [x] Categorize the PR by adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog.
- [x] Link this PR to related issues or PRs.

### CODE REVIEW

File: component/component/vshn_redis.jsonnet [hunk 1]
```diff
@@ -93,6 +93,25 @@ local restoreRole = kube.ClusterRole(restoreRoleName) {
     },
   ],
 };
+local helmMonitoringClusterRoleName = 'allow-helm-monitoring-resources';
+local helmMonitoringClusterRole = kube.ClusterRole(helmMonitoringClusterRoleName) {
```
Comment (zugao) [inline component/component/vshn_redis.jsonnet:97]: You can remove the variable defined above and just pass the name here.

File: component/component/vshn_redis.jsonnet [hunk 2]
```diff
@@ -377,6 +396,16 @@ local composition =
...
           },
           values: {
+            metrics: {
+              enabled: true,
+              containerSecurityContext: {
+                enabled: false,
+              },
+              serviceMonitor: {
+                enabled: true,
+                namespace: '',  // patched
```
Comment (zugao) [inline component/component/vshn_redis.jsonnet:406]: Where this is being patched?

File: component/component/vshn_redis.jsonnet [hunk 3]
```diff
@@ -377,6 +396,16 @@ local composition =
             version: redisParams.helmChartVersion,
           },
           values: {
+            metrics: {
+              enabled: true,
+              containerSecurityContext: {
+                enabled: false,
```
Comment (zugao) [inline component/component/vshn_redis.jsonnet:402]: This I suppose is to make it work on OpenShift?

**Review by zugao — COMMENTED** [review]
Looks good overall but there are things to take a look and change and please remove `compiled` `dependencies` and `vendor` directories from this PR.

File: component/component/vshn_redis.jsonnet (hunk 2 above)
Comment (wejdross) [reply component/component/vshn_redis.jsonnet:406]: in line 645: 
```
comp.FromCompositeFieldPathWithTransformPrefix('metadata.labels[crossplane.io/composite]', 'spec.forProvider.values.metrics.serviceMonitor.namespace', 'vshn-redis'),
```

Comment (wejdross) [reply component/component/vshn_redis.jsonnet:406]: in line 645: 
```
comp.FromCompositeFieldPathWithTransformPrefix('metadata.labels[crossplane.io/composite]', 'spec.forProvider.values.metrics.serviceMonitor.namespace', 'vshn-redis'),
```

File: component/component/vshn_redis.jsonnet (hunk 3 above)
Comment (wejdross) [reply component/component/vshn_redis.jsonnet:402]: exactly

Comment (zugao) [reply component/component/vshn_redis.jsonnet:402]: Don't we have a variable to figure out if the cluster has distribution OpenShift? We want our AppCat to run also on on OpenShift clusters. Like [here](https://github.com/vshn/component-appcat/blob/master/component/component/vshn_redis.jsonnet#L37).

File: component/component/vshn_redis.jsonnet [hunk 4]
```diff
@@ -94,6 +94,25 @@ local restoreRole = kube.ClusterRole(restoreRoleName) {
   ],
 };
 
+local helmMonitoringClusterRole = kube.ClusterRole('allow-helm-monitoring-resources') {
+  rules: [
+    {
+      apiGroups: [ 'monitoring.coreos.com' ],
+      resources: [ 'servicemonitors' ],
+      verbs: [ '*' ],
```
Comment (zugao) [inline component/component/vshn_redis.jsonnet:102]: I was thinking about this configuration, are we sure we need all verbs? Maybe we need only `list, watch and get` only.

Comment (wejdross) [reply component/component/vshn_redis.jsonnet:102]: it also requires create, update and delete, therefore I used `'*'` asteriks value - easier to read :)

**Pull request merged**